### PR TITLE
Readme todos

### PR DESCRIPTION
**Pull request by marcinkp**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] My code follows the [code style](/CONTRIBUTING.md#coding-conventions) of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

---
I hereby agree to the terms of the Knot.x Contributor License Agreement.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -104,16 +104,101 @@ The data source parameters can be also configured in Fragment and merged with de
...
-## Data Source Caching
-// TODO
+## Data Source Calls Caching
 
+Template might consists of more than one Data Source Adapter call. 
+It's also possible that there are multiple fragments on the page, each using same call. 
+Knot.x Data Bridge does caching results of Adapter calls to avoid multiple calls for the same data. 
+Caching is performed within page request scope, this means another request will not get cached data.
 
+<span id="data-source-adapters"></span>
```
Comment (tomaszmichalak) [inline README.md]: Required?

File: README.md [hunk 2]
```diff
@@ -104,16 +104,101 @@ The data source parameters can be also configured in Fragment and merged with de
...
 source definition based on a namespace.
 
-## Data Source Caching
-// TODO
+## Data Source Calls Caching
 
+Template might consists of more than one Data Source Adapter call. 
+It's also possible that there are multiple fragments on the page, each using same call. 
+Knot.x Data Bridge does caching results of Adapter calls to avoid multiple calls for the same data. 
+Caching is performed within page request scope, this means another request will not get cached data.
```
Comment (tomaszmichalak) [inline README.md]: Please add some information how we discover that we have the same data source calls.

File: README.md [hunk 3]
```diff
@@ -104,16 +104,101 @@ The data source parameters can be also configured in Fragment and merged with de
...
 
+<span id="data-source-adapters"></span>
 ## Data Sources Adapters
-// TODO
+Data Sources Adapter is Component of a system, that mediate communication between Knot.x Data Bridge Knot and external 
+services that deliver data injected into template. In short, Data Sources Adapter acts as a element
```
Comment (tomaszmichalak) [inline README.md]: Data Bridge adds data into Knot Context.

File: README.md [hunk 4]
```diff
@@ -104,16 +104,101 @@ The data source parameters can be also configured in Fragment and merged with de
...
+Data Sources Adapter accepts message of type `DataSourceAdapterRequest` with the following data:
 
+ - request - object with all data of an original client http request,
+ - params - all additional params defined in configuration or passed in template.
+ 
+Result generated by Adapter must be of type `DataSourceAdapterResponse` with `clientResponse` field set:
+
+ - body - field  is supposed to carry on the actual response 
+ from the service in JSON format. Can be a single object or an array of objects 
+ - status code - http status code of processing (?)
```
Comment (tomaszmichalak) [inline README.md]: Please update this section.

File: README.md (hunk 1 above)
Comment (marcinkp) [reply README.md]: (#Data Sources Adapters) doesn't work for me :(

File: README.md (hunk 2 above)
Comment (marcinkp) [reply README.md]: added

File: README.md (hunk 3 above)
Comment (marcinkp) [reply README.md]: That's true, but a user/ developer expect data available in templete Knotx Context is not accessible from markup

File: README.md (hunk 4 above)
Comment (marcinkp) [reply README.md]: sure. it was before I started working on fallbacks - before status code wasnt never used by Knotx Bridge

**Pull request closed without merging**